### PR TITLE
Seal fail logic; refactoring

### LINKED ISSUE #6764: PEARL Pressure Controller: Detect Seal Fail DB Logic
As a: developer

I would like to add DB logic to detect if a seal failure has occurred to trigger a seal fail mode which can only be turned off once the device has been power cycled as specified in the manual. 

The device allows for a seal fail value to be set by the user. 
When the pressure suddenly drops by the seal fail value set, a seal fail mode should be set which locks out any other functionality until power cycled.

When a seal fail mode is active, this should be displayed using an LED in the OPI. 

### Acceptance Criteria
What is the acceptance criteria?
- [ ] Seal fail mode initiated if pressure drops suddenly by seal fail value
- [ ] Seal fail mode displayed in OPI as LED

### Extra Information
_Why it is needed?_
Detecting a seal failure is an important requirement of the device in order to function as expected.

_Where required files/links are_
- Manual: `\\isis\shares\ISIS_Experiment_Controls\Manuals\PEARL Pressure Controller\PEARL handbook_v3_5.docx`
- Labview Vi: `C:\LabVIEW Modules\Instruments\PEARL\PEARL Pressure Cell Controller`
- Please see conversation in #6040 to view a screenshot of LabView VI for reference.
- PEARL Pressure Controller Emulator: `C:\Instrument\Apps\EPICS\support\PearlPressureController\master`
- PEARL Pressure Controller IOC: `C:\Instrument\Apps\EPICS\ioc\master\PEARLPC`
- PEARL Pressure Controller OPI: `C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources`

### LINKED ISSUE #6768: PEARL Pressure Controller:  Joy Buttons Active
As a: developer

I would like to implement the DB logic and update OPI to include an LED to display whether or not the joy buttons are active to reflect the existing functionality present within the LabView VI based on information which will be documented as part of #6763 as the manual does not provide sufficient information to achieve desired functionality.

### Acceptance Criteria
What is the acceptance criteria?
- [ ] DB logic to control joy buttons active LED
- [ ] LED in OPI to display whether or not joy buttons are active

### Extra Information
_Why it is needed?_
To replicate existing functionality present within the LabView VI

_Where required files/links are_
- Manual: `\\isis\shares\ISIS_Experiment_Controls\Manuals\PEARL Pressure Controller\PEARL handbook_v3_5.docx`
- Labview Vi: `C:\LabVIEW Modules\Instruments\PEARL\PEARL Pressure Cell Controller`
- Please see conversation in #6040 to view a screenshot of LabView VI for reference.
- PEARL Pressure Controller Emulator: `C:\Instrument\Apps\EPICS\support\PearlPressureController\master`
- PEARL Pressure Controller IOC: `C:\Instrument\Apps\EPICS\ioc\master\PEARLPC`
- PEARL Pressure Controller OPI: `C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources`

### LINKED ISSUE #6779: PEARL Pressure Controller: Cyl A and B Upper and Lower
As a: developer

I would like to create the DB logic and corresponding fields in the OPI to display cylinder A and B upper and lower values using information documented as part of #6763 to match existing functionality present in the LabVIEW VI.

Cylinder A and B upper and lower values are not documented verbosely in the manual, describing only that they can be displayed on the pressure controllers front panel display with no documentation describing how to access the values directly to be used in the OPI similarly to the LabVIEW VI implementation.

### Acceptance Criteria
- [ ] Create DB logic to read back cylinder A and B upper and lower values.
- [ ] Create OPI fields to display cylinder A and B upper and lower values.

### Extra Information
_Why it is needed?_
Functionality exists in LabView VI and should be replicated in IBEX GUI device screen for consistency.

_Where required files/links are_
- Manual: `\\isis\shares\ISIS_Experiment_Controls\Manuals\PEARL Pressure Controller\PEARL handbook_v3_5.docx`
- Labview Vi: `C:\LabVIEW Modules\Instruments\PEARL\PEARL Pressure Cell Controller`
- Please see conversation in #6040 to view a screenshot of LabView VI for reference.
- PEARL Pressure Controller Emulator: `C:\Instrument\Apps\EPICS\support\PearlPressureController\master`
- PEARL Pressure Controller IOC: `C:\Instrument\Apps\EPICS\ioc\master\PEARLPC`
- PEARL Pressure Controller OPI: `C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources`

### LINKED ISSUE #7044: Pearl Pressure Controller: Umbrella Issue For Cell and Pump Pressure
## Where?
_Where is the issue likely to be (be as specific as possible e.g. filepaths)_

This issue is an umbrella issue to complete the below linked issues as part of development for the PEARL Pressure Controller for Wish ready for the next cycle.
- [#6774](https://github.com/ISISComputingGroup/IBEX/issues/6774)
- [#6770](https://github.com/ISISComputingGroup/IBEX/issues/6770)

## How?
_How did the issue come about/known cause of issue if any? (**delete subheader if not applicable**)_

How did the issue come about/known cause of issue if any? (delete subheader if not applicable)
WISH require the Pearl Pressure controller for the next cycle within IBEX. To make this possible, an IOC and OPI need to be developed to match existing functionality present in the LabVIEW VI implementation of the Pearl Pressure controller.

## Acceptance criteria 
- [ ] All associated issues have been completed
- [ ] All associated issues have been tested.

## How to Test 
_verbose instructions for reviewer to test changes
(**Add before making a PR**)_

### Extra Information
_Why it is needed?_
- Functionality exists in LabView VI and should be replicated in IBEX GUI device screen for consistency.

_Where required files/links are_
- Manual: `\\isis\shares\ISIS_Experiment_Controls\Manuals\PEARL Pressure Controller\PEARL handbook_v3_5.docx`
- Labview Vi: `C:\LabVIEW Modules\Instruments\PEARL\PEARL Pressure Cell Controller`
- Please see conversation in #6040 to view a screenshot of LabView VI for reference.
- PEARL Pressure Controller Emulator: `C:\Instrument\Apps\EPICS\support\PearlPressureController\master`
- PEARL Pressure Controller IOC: `C:\Instrument\Apps\EPICS\ioc\master\PEARLPC`
- PEARL Pressure Controller OPI: `C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources`

### LINKED ISSUE #7193: Pearl Pressure controller: implement readbacks from `ls` command (sample protection settings)
Maybe in an advanced screen

As discussed with @FreddieAkeroyd and @Adam-Szw in a meeting 20/06/2022

### PR DESCRIPTION
**Pull request by Tom-Willemsen**
- Fixes https://github.com/ISISComputingGroup/IBEX/issues/6764
- Fixes https://github.com/ISISComputingGroup/IBEX/issues/7193
- Fixes https://github.com/ISISComputingGroup/IBEX/issues/7044
- Fixes https://github.com/ISISComputingGroup/IBEX/issues/6779
- Fixes https://github.com/ISISComputingGroup/IBEX/issues/6768

### CODE REVIEW

File: PearlPCSup/devPearl.db [hunk 1]
```diff
@@ -147,28 +157,123 @@ record(longin, "$(P)MN_PRESSURE"){
...
+
+record(calc, "$(P)INPUTS:_RAW") {
+    # .A written to from protocol
+    field(INPB, "$(P)STATUS_ARRAY CP MS")
+	field(CALC, "A")
+}
+
+record(mbbiDirect, "$(P)INPUTS") {
+    field(INP, "$(P)INPUTS:_RAW CP MS")
+	# B8 = Emergency Stopped released
```
Comment (FreddieAkeroyd) [inline PearlPCSup/devPearl.db:196]: Do you need to use an intermediate calc record - can you not read directly into an mbbiDirect from the record redirection?

File: PearlPCSup/PearlPC.proto [hunk 2]
```diff
@@ -78,5 +101,26 @@ get_memory{
...
+	wait 100;
+}
+
+set_pos_lim {
+    out "d+%#04d";
+	wait 100;
+}
+
+set_neg_lim {
+    out "d-%#04d";
```
Comment (FreddieAkeroyd) [inline PearlPCSup/PearlPC.proto:114]: I think this may be `%#01d`

File: PearlPCSup/devPearl.db (hunk 1 above)
Comment (Tom-Willemsen) [reply PearlPCSup/devPearl.db:196]: The calc is to ensure disconnection alarms propagate to this record, as I don't think those would be propagated with just a pure redirection?

File: PearlPCSup/PearlPC.proto (hunk 2 above)
Comment (Tom-Willemsen) [reply PearlPCSup/PearlPC.proto:114]: I think you're right for the `o+` and `o-` commands, I've changed that in those commands. For `d+` and `d-` I think 4 digits is correct according to the manual.

File: PearlPCSup/devPearl.db (hunk 1 above)
Comment (FreddieAkeroyd) [reply PearlPCSup/devPearl.db:196]: Ah i see, via the `INPB`. I am not sure what would happen with redirection on disconnection

Comment (Tom-Willemsen) [reply PearlPCSup/devPearl.db:196]: I think it never notices any alarm - which is why we have this pattern elsewhere (sometimes we use it via [`error_setter.template`](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IOC-Utilities#error-setting), but in this case I thought it was simple enough not to need that template). I'm happy to use the template if you prefer though.

**Pull request merged**